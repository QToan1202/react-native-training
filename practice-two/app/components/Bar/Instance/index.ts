export { default as HomeBar } from './Home'
export { default as BrowseBar } from './Browse'
export { default as CategoryBar } from './Category'
export { default as BackBar } from './Checkout'
export { default as TabBar } from './Tabs'
