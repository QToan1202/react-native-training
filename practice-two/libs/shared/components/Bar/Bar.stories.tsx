import { ComponentMeta, ComponentStory } from '@storybook/react-native'
import { action } from '@storybook/addon-actions'

import { Bar, Button, IconButton, Search } from '../index'

export default {
  title: 'components/Bar',
  component: Bar,
} as ComponentMeta<typeof Bar>

const IconList = <IconButton icon={require('@assets/favicon.png')} onPress={action('press')} />

const Template: ComponentStory<typeof Bar> = (args) => <Bar {...args} />

export const Default = Template.bind({})
Default.args = {
  title: 'Title',
  align: 'flex-start',
}

export const BarWithButton = Template.bind({})
BarWithButton.args = {
  title: 'Heading',
  align: 'space-between',
  IconList,
  children: (
    <>
      <Button
        onPress={action('button-press')}
        title="sort by"
        variant="tertiary"
        shrink
        style={{ flex: 1 }}
        leftIcon={require('@assets/favicon.png')}
      />
      <Button
        onPress={action('button-press')}
        title="location"
        variant="tertiary"
        shrink
        style={{ flex: 1 }}
        leftIcon={require('@assets/favicon.png')}
      />
      <Button
        onPress={action('button-press')}
        title="category"
        variant="tertiary"
        shrink
        style={{ flex: 1 }}
        leftIcon={require('@assets/favicon.png')}
      />
    </>
  ),
}

export const BarWithSearch = Template.bind({})
BarWithSearch.args = {
  title: 'Search',
  align: 'space-between',
  children: <Search placeholder="Search Product" />,
}
