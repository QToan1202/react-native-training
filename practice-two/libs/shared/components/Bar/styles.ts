import { StyleSheet } from 'react-native'

import { COLORS } from '../../constants'

const styles = StyleSheet.create({
  bar: {
    backgroundColor: COLORS.PRIMARY,
    rowGap: 24,
  },
  title: {
    textTransform: 'capitalize',
  },
  backBtn: {
    position: 'absolute',
    left: 0,
    top: 0,
    zIndex: 1,
  },
  itemsContainer: {
    justifyContent: 'space-between',
  },
})

export default styles
