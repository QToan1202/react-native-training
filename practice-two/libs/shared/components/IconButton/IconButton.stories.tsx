import { ComponentMeta, ComponentStory } from '@storybook/react-native'
import { action } from '@storybook/addon-actions'

import { IconButton } from '../index'

export default {
  title: 'components/IconButton',
  component: IconButton,
} as ComponentMeta<typeof IconButton>

const Template: ComponentStory<typeof IconButton> = (args) => <IconButton {...args} />

export const Default = Template.bind({})
Default.args = {
  icon: require('@assets/favicon.png'),
  accessibilityLabel: 'like button',
  onPress: action('Pressing'),
}

export const ButtonWithoutBg = Template.bind({})
ButtonWithoutBg.args = {
  icon: require('@assets/favicon.png'),
  noBackground: true,
  onPress: action('Pressing'),
}
