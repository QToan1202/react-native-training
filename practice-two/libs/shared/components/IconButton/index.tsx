import { memo } from 'react'
import isEqual from 'react-fast-compare'

import StyledIconButton, { CustomButtonProps } from './styles'

export type IconButtonProps = CustomButtonProps

const IconButton = ({ children, ...rest }: IconButtonProps) => {
  return <StyledIconButton {...rest}>{children}</StyledIconButton>
}

export default memo(IconButton, isEqual)
