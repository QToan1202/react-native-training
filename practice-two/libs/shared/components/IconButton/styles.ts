import { Circle, GetProps, styled } from 'tamagui'

const CustomButton = styled(Circle, {
  name: 'IconButton',
  alignSelf: 'flex-start',
  padding: '$space.2',
  backgroundColor: '$color.blur_white',
  pressStyle: {
    opacity: 0.6,
  },

  variants: {
    noBackground: {
      true: {
        backgroundColor: '$color.transparent',
      },
    },
  } as const,
  defaultVariants: {
    noBackground: false,
  },
})

export type CustomButtonProps = GetProps<typeof CustomButton>

export default CustomButton
