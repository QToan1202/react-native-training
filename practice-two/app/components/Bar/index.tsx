import React, { ReactNode, memo } from 'react'
import isEqual from 'react-fast-compare'
import { useSafeAreaInsets } from 'react-native-safe-area-context'
import { XStack, XStackProps, YStack, YStackProps } from 'tamagui'

import Heading from '@components/Heading'
import IconButton from '@components/IconButton'

import styles from './styles'

export type BarProps = {
  title?: string
  showBackBtn?: boolean
  align?: XStackProps['justifyContent']
  IconList?: ReactNode
  children?: ReactNode
  onPressBack?: () => void
} & YStackProps

const Bar = ({
  title,
  showBackBtn = false,
  align = 'flex-start',
  IconList,
  children,
  onPressBack,
  ...rest
}: BarProps) => {
  const checkProps = title || showBackBtn || IconList
  const insets = useSafeAreaInsets()

  return (
    <YStack
      paddingHorizontal={16}
      paddingTop={insets.top}
      paddingBottom={22}
      backgroundColor="$color.primary"
      space="$space.5"
      {...rest}
    >
      {checkProps && (
        <XStack space={4} alignItems="center" justifyContent={align}>
          {showBackBtn && (
            <IconButton
              style={styles.backBtn}
              icon={require('@assets/back.png')}
              onPress={onPressBack}
              noBackground
            />
          )}
          {title && <Heading content={title} textTransform="capitalize" />}
          {IconList && (
            <XStack space={4} alignItems="center">
              {IconList}
            </XStack>
          )}
        </XStack>
      )}
      {children && (
        <XStack space={4} alignItems="center" justifyContent="space-between">
          {children}
        </XStack>
      )}
    </YStack>
  )
}

export default memo(Bar, isEqual)
