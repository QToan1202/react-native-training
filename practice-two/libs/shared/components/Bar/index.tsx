import React, { ReactNode, memo } from 'react'
import isEqual from 'react-fast-compare'
import { useSafeAreaInsets } from 'react-native-safe-area-context'
import { XStack, XStackProps, YStack, YStackProps } from 'tamagui'

import BackIcon from '@assets/back.svg'

import Heading from '../Heading'
import IconButton from '../IconButton'

import styles from './styles'

export type BarProps = {
  title?: string
  showBackBtn?: boolean
  align?: XStackProps['justifyContent']
  IconList?: ReactNode
  children?: ReactNode
  onPressBack?: () => void
} & YStackProps

const Bar = ({
  title,
  showBackBtn = false,
  align = 'flex-start',
  IconList,
  children,
  onPressBack,
  ...rest
}: BarProps) => {
  const checkProps = title || showBackBtn || IconList
  const insets = useSafeAreaInsets()

  return (
    <YStack
      paddingHorizontal={16}
      paddingTop={insets.top}
      paddingBottom={22}
      backgroundColor="$color.primary"
      space="$space.5"
      {...rest}
    >
      {checkProps && (
        <XStack space={4} alignItems="center" justifyContent={align}>
          {showBackBtn && (
            <IconButton style={styles.backBtn} onPress={onPressBack} noBackground>
              <BackIcon />
            </IconButton>
          )}
          {title && <Heading content={title} textTransform="capitalize" />}
          {IconList && (
            <XStack space={4} alignItems="center">
              {IconList}
            </XStack>
          )}
        </XStack>
      )}
      {children && (
        <XStack space={4} alignItems="center" justifyContent="space-between">
          {children}
        </XStack>
      )}
    </YStack>
  )
}

export default memo(Bar, isEqual)
