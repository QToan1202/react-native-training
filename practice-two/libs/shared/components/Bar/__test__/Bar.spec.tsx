import { render, screen } from '@testing-library/react-native'

import { COLORS } from '@constants'
import IconButton from '@components/IconButton'
import { IIconList } from '@types'

import Bar from '../index'

const iconList: IIconList[] = [
  {
    label: 'favicon',
    icon: require('@assets/favicon.png'),
  },
]
const mockPressBackFn = jest.fn()

describe('Testing Bar component', () => {
  beforeEach(() => {
    render(
      <Bar
        title="heading"
        IconList={iconList.map(({ icon, label }) => (
          <IconButton key={label} icon={icon} />
        ))}
        showBackBtn
        onPressBack={mockPressBackFn}
        testID="bar"
      />
    )
  })

  it('Render Bar component success', () => {
    const component = render(<Bar title="home" IconList={[]} />).toJSON()

    expect(component).toMatchSnapshot()
  })

  it('Bar component render with primary color', () => {
    const component = screen.getByTestId('bar')

    expect(component).toHaveStyle({ backgroundColor: COLORS.PRIMARY })
  })

  it('Bar component contain text component that match with title prop', () => {
    const bar = screen.getByTestId('bar')
    const title = screen.getByText('heading')

    expect(bar).toContainElement(title)
  })
})
