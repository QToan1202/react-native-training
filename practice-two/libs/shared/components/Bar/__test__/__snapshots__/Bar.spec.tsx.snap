// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Bar component Render Bar component success 1`] = `
<View
  style={
    [
      {
        "paddingHorizontal": 16,
        "paddingVertical": 22,
      },
      {
        "backgroundColor": "#33907C",
        "rowGap": 24,
      },
      {
        "paddingTop": 0,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "columnGap": 4,
          "flexDirection": "row",
        },
        {
          "justifyContent": "flex-start",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#fff",
            "fontFamily": "Montserrat_700Bold",
            "fontSize": 24,
          },
          false,
          {
            "textTransform": "capitalize",
          },
        ]
      }
    >
      home
    </Text>
    <View
      style={
        {
          "alignItems": "center",
          "columnGap": 4,
          "flexDirection": "row",
        }
      }
    />
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "columnGap": 4,
          "flexDirection": "row",
        },
        {
          "justifyContent": "space-between",
        },
      ]
    }
  />
</View>
`;
